packages/service-library/src/servicelib/redis_utils.py

```diff
@@ -24,12 +30,32 @@ def exclusive(
     redis: the redis client SDK
     lock_key: a string as the name of the lock (good practice: app_name:lock_name)
     lock_value: some additional data that can be retrieved by another client
+
+    Raises:
+        - ValueError if used incorrectly
+        - CouldNotAcquireLockError if the lock could not be acquired
     """
 
-    def decorator(func):
+    if not lock_key:
+        msg = "lock_key cannot be empty string!"
+        raise ValueError(msg)
+
+    def decorator(func: Callable[P, Awaitable[R]]) -> Callable[P, Awaitable[R]]:
         @functools.wraps(func)
-        async def wrapper(*args, **kwargs):
-            async with redis.lock_context(lock_key=lock_key, lock_value=lock_value):
+        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+            redis_lock_key = lock_key
+            if callable(lock_key):
+                redis_lock_key = lock_key(*args, **kwargs)
```

Comment:
Maybe I am just confused, but is it intended that you evaluate the `str | bytes | None` `lock_key` on args and kwargs? I would have expected you would get some redis object first and evaluate.

Reply:
hmm not sure I understand here.
the `lock_key` is either a callable returning a string or a string.
Actually fixed the typing, hope that makes it clearer.